flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/utils/TimeTestUtil.scala

```diff
@@ -52,14 +56,47 @@ object TimeTestUtil {
     }
   }
 
+  /**
+   * A streaming operator to emit records and watermark depends on the input data.
+   * The last emitted watermark will be stored in state to emit it again on recovery.
+   * This is necessary for late arrival testing with [[FailingCollectionSource]].
+   */
   class EventTimeProcessOperator[T]
     extends AbstractStreamOperator[T]
       with OneInputStreamOperator[Either[(Long, T), Long], T] {
 
+    private var currentWatermark: Long = 0L
+    private var watermarkState: ListState[JLong] = _
+
+    override def snapshotState(context: StateSnapshotContext): Unit = {
+      super.snapshotState(context)
+      watermarkState.clear()
+      watermarkState.add(currentWatermark)
+    }
+
+    override def initializeState(context: StateInitializationContext): Unit = {
+      super.initializeState(context)
+      // use union list state to get the max watermark
+      watermarkState = context.getOperatorStateStore.getUnionListState(
```

Comment:
What's the idea of using max watermark?

In the used test cases, the sources have only a DOP of 1 (at least from what I have seen). So it seems more complicated than necessary as the union state is actually not used with more than one element afaik.

Even if we have a DOP>1, it might even be harmful. Let's say we have two source subtask. By using the max watermark on all sources, we might declare some elements late, even though they are not, because we used the watermark of the other source subtasks.

I also haven't seen any indication that this class is used for rescaling. So union state might be over-engineered.

Reply:
Thanks for pointing it out. I agree with you it's over-engineered. It is not simple in rescale case. I will use simple list state as we won't rescale in tests.